File: easy/valid-parentheses.ts
// https://leetcode.com/problems/valid-parentheses/submissions/

function isValid(s: string): boolean {
    const stack: string[] = [];
    const dict: {[key: string]: string} = {
        '(': ')',
        '[': ']',
        '{': '}'
    }

    for (let p of s) {
        if (dict[p]) {
            stack.push(p)
        } else {
            let lastPar = stack.pop()
            if (!lastPar) return false;

            if (dict[lastPar] === p) {
                continue;
            } else {
                return false;
            }
        }
    }

    return !stack.length;
};

console.assert(isValid("()[]{}"))
console.assert(isValid(""))
console.assert(isValid("({{[]}})(") === false)
console.assert(isValid("(") === false)

File: ds_1/day_9/valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses/


from collections import deque


class Solution(object):
    def isValid(self, s):
        stack = deque()

        d = {
            '(': ')',
            '{': '}',
            '[': ']'
        }

        for i in s:
            if i in d:
                stack.append(i)
            else:
                if len(stack) == 0:
                    return False
                p = stack.pop()
                if i != d[p]:
                    return False

        return len(stack) == 0